File: octave/ecuaciones_no_lineales/codigos/newton_raphson.m
function [x, k, error]=newton_raphson(f, xo, tol, imax) 
    pkg load symbolic 
    syms x;

    % forma 1 %
    %f1 = str2func(f);
    %df = diff(sym(f1))

    % forma 2 %
    f1 = matlabFunction(sym(f)); 
    df = matlabFunction(diff(sym(f1)));
    k = 0;
    error = tol + 1;
    e = [];
    while (error > tol && k < imax)
        deno = df(xo);
        if (deno == 0)
            x = [];
            k=[]; 
            error=[];
            disp('La función se indefine.')
        else
            xn = xo - (f1(xo)/deno);
            error = abs(xo - (f1(xo)/deno));
            e=[e error];
            xo = xn;
            k = k + 1;
        end
    end
    plot(1:k,e)
    xlabel('iter (k)')
    ylabel('Error (|f(x_k)|)')
    title('Error del Metodo de la Newton-Raphson')
end

% forma 1 %
%[x,k,error]=newton_raphson('@(x)cos(2*x).^2-x.^2',3/4,10^-8,10)

% forma 2 %
%[x,k,error]=newton_raphson(cos(2*x).^2-x.^2,3/4,10^-8,10)

File: python/ecuaciones_no_lineales/newton_raphson.py
import math
import matplotlib.pyplot
from sympy import *


def newton(f, xo, tol, iterMax):
    x = Symbol('x')
    f1 = sympify(f)
    df = diff(f1)
    error = tol + 1
    k = 0
    it = []
    e = []

    while tol < error and k < iterMax:
        if df.evalf(subs={x: xo}) == 0:
            x = []
            k = []
            error = []
            print("La funcion se indefine en xo.")
        else:
            xn = xo - (f1.evalf(subs={x: xo}) / df.evalf(subs={x: xo}))
            error = abs(xn)
            it.append(k)
            e.append(error)
            xo = xn

        k += 1
    matplotlib.pyplot.plot(it, e)
    ans = {
        "x": xn,
        "k": k,
        "error": e[-1]
    }
    return ans

# newton('cos(2*x)**2-x**2',3/4,10**-8,10)
